keps/sig-node/2238-liveness-probe-grace-period/README.md

```diff
@@ -0,0 +1,569 @@
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Liveness probes currently use the `terminationGracePeriodSeconds` on both
+normal shutdown and when probes fail. Hence, if a long termination period is
+set, and a liveness probe fails, a workload will not be promptly restarted
+because it will wait for the full termination period.
+
+I propose adding a new field to liveness probes, `gracePeriodSeconds`. When
+set, it will override the `terminationGracePeriodSeconds` for liveness
+termination. This maintains the current behaviour if desired while providing
+configuration to address this unintended behaviour.
+
+## Motivation
+
+This change is important to give users the option to decouple these fields, as
+the current behaviour is counter-intuitive and doesn't provide an option to
+decouple.
+
+[Other implementation options](#alternatives) might avoid an API change, but
+would end up breaking backwards compatibility which some users may have come to
+[expect/rely on](https://github.com/kubernetes/kubernetes/issues/64715#issuecomment-756100727),
+even if we take a phased approach to introduce them.
+
+An example of a scenario where this bug was apparent is a long outage reported
+by an ingress controller:
+
+> the controller uses a 3600s grace period so that it can drain long requests
+> from end user app traffic (it listens on endpoints that can remain available
+> for hours like host network, but also incoming tcp load balancers which have
+> open connections), but when the controller wedges it takes an hour to
+> resolve, which is the worst possible outcome (the process wedges immediately,
+> but kube has to wait the full hour to kill it, defeating the goal of
+> liveness). [comment](https://github.com/kubernetes/kubernetes/issues/64715#issuecomment-756201502)
+
+### Goals
+
+- Liveness probes can have a configurable timeout on failure.
+- Existing behaviour is not changed.
+
+### Non-Goals
+
+Any changes to probe behaviour outside of the scope of addressing the
+[bug](https://github.com/kubernetes/kubernetes/issues/64715) this is intended
+to address.
+
+## Proposal
+
+We can add a `terminationGracePeriodSeconds` integer field to the Probe API.
+This field would specify how long the kubelet should wait before forcibly
+terminating the container when a probe fails, and override the pod-level
+`terminationGracePeriodSeconds` value.
```

Comment:
would it be confusing to have value `terminationGracePeriodSeconds` specified in the probe? Example given in the beginning where `terminationGracePeriodSeconds` is used to drain nodes works fine for this new field. What if the `terminationGracePeriodSeconds` is used to make sure that the state is successfully recorded in some database on exit and designed by app developer while liveness probe is designed by "operations". It may be more logical to see this new field on a top level next to the exiting field. Maybe list this into the alternative considerations with explanaition why the new field is under the probe, not on the top level.